### ISSUE TITLE
As a CD automation service account or developer I need to completely delete/recreate the Anthos cluster and associated solution projects from the target organization - in step

### ISSUE THREAD

**Issue by fmichaelobrien** [GoogleCloudPlatform/pubsec-declarative-toolkit#91]
use case: full 2 part CD automation of the KCC cluster and the lz solution - with full tear down of everything or just the lz solution
- action: add a delete action to the arete cli
- this case is for a complete reuse of the org - back to a clean state (minus required roles)

Indirect collaboration on:  https://github.com/GoogleCloudPlatform/pbmm-on-gcp-onboarding/issues/166

Notes:
# removing the solution via arete
- pending (discussions with Shaun and Chris)
your prompts do work well - I like them in the create
just thinking automated CD - will need a cli double minus flag for scripts
I would target an addition (where we run 2+ solutions) - thinking your option 2 (from the cluster)
https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit/tree/main/solutions/landing-zone

create
arete create landing-zone-controller --region=$REGION

delete (option 2)
arete delete landing-zone-controller --region=$REGION
# removing the solution via kubernetes cli (thanks Chris for reminding me to think granular k8s again)
- stick to this for now
```
admin_root@cloudshell:~/wse_github/20220831/pubsec-declarative-toolkit (pubsec-declarative-toolkit-cno)$ export REGION=northamerica-northeast1
admin_root@cloudshell:~/wse_github/20220831/pubsec-declarative-toolkit (pubsec-declarative-toolkit-cno)$ arete create landing-zone-controller --region=$REGION
8:41PM INF Project name will be set to: landing-zone-controller-1pw8k


```
# removing the solution via kpt cli

# removing the solution via gcloud cli

# removing the solution manually
- all projects under pdt
<img width="741" alt="Screen Shot 2022-09-02 at 2 24 19 PM" src="https://user-images.githubusercontent.com/94715080/188215147-527a7165-64fa-4e19-a6e0-44b73958d526.png">

check arete cache - pending on delete command
```
admin_root@cloudshell:~$ ls -la .arete/
total 20
drwxr--r--  2 admin_root admin_root 4096 Sep  2 18:15 .
drwxr-xr-x 11 admin_root admin_root 4096 Sep  1 19:27 ..
-rw-r--r--  1 admin_root admin_root   46 Aug 31 15:06 config.yaml
-rw-------  1 admin_root admin_root  100 Aug 31 15:32 .create
-rw-r--r--  1 admin_root admin_root 1318 Sep  2 18:15 solutions.yaml
admin_root@cloudshell:~$ cat .arete/solutions.yaml
solutions:
    - solution: guardrails
      description: |
        Implementation of the GC Cloud Guardrails Checks. More Info: https://github.com/canada-ca/cloud-guardrails-gcp
      url: https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit.git/solutions/guardrails@main
    - solution: guardrails-policy-bundle
      description: |
        Policy Bundle to help analyze compliance for Guardrails
      url: https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit.git/solutions/guardrails-policies@main
    - solution: kcc-namespaces
      description: |
        Simplified declarative multi-tenancy with project namespaces taken from: https://cloud.google.com/anthos-config-management/docs/tutorials/project-namespace-blueprint
      url: https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit.git/solutions/kcc-namespaces@main
    - solution: sandbox-gke
      description: |
        A private GKE cluster with so many bells and whistles!
      url: https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit.git/solutions/sandbox-gke@main
    - solution: landing-zone
      description: |
        This is a reimplementation of pbmm-on-gcp-onboarding Landing Zone using KRM.
      url: https://github.com/GoogleCloudPlatform/gcp-pbmm-sandbox.git/solutions/landing-zone@main
admin_root@cloudshell:~$ cat .arete/config.yaml
cache: /home/admin_root/.arete
verbose: false
admin_root@cloudshell:~$ cat .arete/.create
steps:
- step: services
- step: network
- step: subnet
- step: config-controller
- step: add-policy
```

- delete the solution project off the target folder
- delete the anthos GKE cluster off the host projectd
- remove the kcc-controller VPC

# Rerun the CC cluster creation and LZ solution
- running into existing .arete config - deleting
```
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-toolkit-cno
2:37PM INF Config Controller setup complete
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ rm -rf
.arete/                .bashrc                .docker/               .npm/                  .redhat/
.bash_history          .cache/                gopath/                .profile               .theia/
.bash_logout           .config/               .kube/                 README-cloudshell.txt  wse_github/
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ ls -la .arete/
total 20
drwxr--r--  2 admin_root admin_root 4096 Sep  2 18:15 .
drwxr-xr-x 12 admin_root admin_root 4096 Sep  6 14:37 ..
-rw-r--r--  1 admin_root admin_root   46 Aug 31 15:06 config.yaml
-rw-------  1 admin_root admin_root  100 Aug 31 15:32 .create
-rw-r--r--  1 admin_root admin_root 1318 Sep  2 18:15 solutions.yaml
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ rm -rf .arete/

admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-toolkit-cno
2:40PM INF Enabling required services...
2:40PM INF Operation "operations/acat.p2-491974186555-2e6beaa9-f3df-4413-9a28-419db485c8e0" finished successfully.
2:41PM INF Creating Config Controller Cluster....
2:41PM FTL  error="ERROR: (gcloud.anthos.config.controller.create) ALREADY_EXISTS: Resource 'projects/pubsec-declarative-toolkit-cno/locations/northamerica-northeast1/krmApiHosts/pdt-cno-kcc' already exists- '@type': type.googleapis.com/google.rpc.ResourceInfo  resourceName: projects/pubsec-declarative-toolkit-cno/locations/northamerica-northeast1/krmApiHosts/pdt-cno-kcc"

deleting project - attempt to reuse may fail on 30 day deleted cache - will try
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ gcloud projects delete pubsec-declarative-toolkit-cno
Your project will be deleted.

Do you want to continue (Y/n)?  y

Deleted [https://cloudresourcemanager.googleapis.com/v1/projects/pubsec-declarative-toolkit-cno].

You can undo this operation for a limited period by running the command below.
    $ gcloud projects undelete pubsec-declarative-toolkit-cno

See https://cloud.google.com/resource-manager/docs/creating-managing-projects for information on shutting down projects.
admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-toolkit-cno
✔ My Billing Account - 019..3D
✔ nuage-cloud.org - 471..7
✔ Folder Level
✔ pdt - 346..8
2:44PM FTL  error="ERROR: (gcloud.projects.create) Project creation failed. The project ID you specified is already in use by another project. Please try an alternative ID."

admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-toolkit-cno2
✔ My Billing Account - 01..3D
✔ nuage-cloud.org - 471924274947
✔ Folder Level
✔ pdt - 346242644868
2:45PM FTL  error="ERROR: (gcloud.projects.create) argument PROJECT_ID: Bad value [pubsec-declarative-toolkit-cno2]: Project IDs are immutable and can be set only during project creation. They must start with a lowercase letter and can have lowercase ASCII letters, digits or hyphens. Project IDs must be between 6 and 30 characters.Usage: gcloud projects create [PROJECT_ID] [optional flags]  optional flags may be  --enable-cloud-apis | --folder | --help | --labels |                         --name | --organization | --set-as-defaultFor detailed information on this command and its flags, run:  gcloud projects create --help"

30 char limit

admin_root@cloudshell:~ (landing-zone-controller-w8hwa)$ arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-tk-cno2
✔ My Billing Account - 019952-0D0AAC-777E3D
✔ nuage-cloud.org - 471924274947
✔ Folder Level
✔ pdt - 346242644868
2:48PM INF Create in progress for [https://cloudresourcemanager.googleapis.com/v1/projects/pubsec-declarative-tk-cno2].Waiting for [operations/cp.7885851846085518239] to finish.....done.Enabling service [cloudapis.googleapis.com] on project [pubsec-declarative-tk-cno2]...Operation "operations/acat.p2-153970848512-8ffc1200-8c5a-42fd-b142-e11cdaf69191" finished successfully.Updated property [core/project] to [pubsec-declarative-tk-cno2].
2:48PM INF Creating Config Controller Cluster....
2:48PM FTL  error="API [krmapihosting.googleapis.com] not enabled on project [153970848512]. Would you like to enable and retry (this will take a few minutes)? (y/N)?  ERROR: (gcloud.anthos.config.controller.create) PERMISSION_DENIED: KRM API Hosting API has not been used in project 153970848512 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/krmapihosting.googleapis.com/overview?project=153970848512 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.- '@type': type.googleapis.com/google.rpc.Help  links:  - description: Google developers console API activation    url: https://console.developers.google.com/apis/api/krmapihosting.googleapis.com/overview?project=153970848512- '@type': type.googleapis.com/google.rpc.ErrorInfo  domain: googleapis.com  metadata:    consumer: projects/153970848512    service: krmapihosting.googleapis.com  reason: SERVICE_DISABLED"

rerun on recently created project - or run on an existing project to avoid the service enablement missing wait timer

                                                      arete create pdt-cno-kcc --region=northamerica-northeast1 --project=pubsec-declarative-tk-cno2
4:39PM INF Enabling required services...
4:40PM INF Operation "operations/acf.p2-153970848512-b3d4a2a6-fe02-4a5b-8f5d-d27d917f6527" finished successfully.
4:40PM INF Creating Network...

........................................................................................done.Created instance [pdt-cno-kcc].Fetching cluster endpoint and auth data.kubeconfig entrgenerated for krmapihost-pdt-cno-kcc.
5:09PM INF Add SA to roles/owner role...
5:09PM INF Config Controller setup complete
```
Reference
- https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit/issues/93
- https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit/issues/94

# deletion using anthos delete
Trying Chris's suggested
```
gcloud anthos config controller delete
```
Or the previous anthos cluster deletion doc
http://wiki.obrienlabs.cloud/display/DEV/Deploying+a+Landing+Zone+on+Google+Cloud#DeployingaLandingZoneonGoogleCloud-Pausingtheanthoscluster
via
http://wiki.obrienlabs.cloud/display/DEV/Deploying+a+Landing+Zone+on+Google+Cloud#DeployingaLandingZoneonGoogleCloud-Pausingtheanthoscluster

reviewing
https://cloud.google.com/anthos-config-management/docs/tutorials/landing-zone#removing_resources

# Temporary CC cluster shutdown and restart

**Comment by cartyc** [comment]
This might not fully answer the question but for CD I would nudge you to using either `kpt` directly from something like cloud build or you're favorite CD tool (working on a skaffold example) or using GitOps via either Git or OCI (docs coming soon). 
Deploying via CD with `arete` is currently out of scope for the tool. 

Is there a preferred method of deployment you would like to see or some gaps in the docs that we should fill in?

**Comment by fmichaelobrien** [comment]
I'll start with cloud build keyed off our workload example https://github.com/GoogleCloudPlatform/pbmm-on-gcp-onboarding/blob/main/docs/google-cloud-landingzone-traffic-generation.md#add-cloud-build-triggers and likely move to cloud deploy and fill in the docs.

For deletion following https://cloud.google.com/anthos-config-management/docs/tutorials/landing-zone#removing_resources
```
admin_root@cloudshell:~$ gcloud config set project pubsec-declarative-tk-cno2
Updated property [core/project].
admin_root@cloudshell:~ (pubsec-declarative-tk-cno2)$ kubectl get nodes
NAME                                                  STATUS   ROLES    AGE    VERSION
gke-krmapihost-pdt-c-krmapihost-pdt-c-23345ad9-6094   Ready    <none>   7h9m   v1.22.11-gke.400
gke-krmapihost-pdt-c-krmapihost-pdt-c-6515adfd-2kgt   Ready    <none>   7h9m   v1.22.11-gke.400
gke-krmapihost-pdt-c-krmapihost-pdt-c-ce2512f1-46tj   Ready    <none>   7h9m   v1.22.11-gke.400
admin_root@cloudshell:~ (pubsec-declarative-tk-cno2)$ gcloud anthos config controller list
NAME: pdt-cno-kcc
LOCATION: northamerica-northeast1
STATE: RUNNING
admin_root@cloudshell:~ (pubsec-declarative-tk-cno2)$ gcloud anthos config controller delete pdt-cno-kcc --location=northamerica-northeast1
You are about to delete instance [pdt-cno-kcc]

Do you want to continue (Y/n)?  y

Delete request issued for: [pdt-cno-kcc]
Waiting for operation [projects/pubsec-declarative-tk-cno2/locations/northamerica-northeast1/operations/operation-1662509152408-5e80b137713f2-3980bef0-55096225] to complete...working.
```

**Comment by shaunmitchellve** [comment]
@fmichaelobrien 

I'm confused on this issue. You mention deleting a solution like the landing zone but then also talk about deleting the KCC cluster? These are two different scenarios and I think needs to be in separate issues.

As @cartyc mentioned, `arete` is not intended to be used in any CD tool chain, this out-of-scope for the cli. The underlying tools that `arete` uses are much better suited for that use-case (for example `kpt`)

**Comment by fmichaelobrien** [comment]
sounds good Shaun, for CD yes ideally I would remove the solution and recreate it - using a trigger off the repo specific to the solutions or particular subfolder.  I mentioned removing the entire GKE cluster (through anthos) as a wider option for changes to the repo in general. 
Yes, I see now that arete is a client tool, I will split this issue and use the advanced anthos level delete/install of the cluster for repo wide CD regression testing

**Comment by fmichaelobrien** [comment]
working out re-create procedure in https://github.com/GoogleCloudPlatform/pubsec-declarative-toolkit/blob/dev/solutions/landing-zone/architecture.md#deleting-the-anthos-cluster

Full scope of this issue is deleting the lz solution first and then deleting the cluster.  There is another issue I will add on just deleting/restarting the cluster and re-associating with the existing lz deployment

**Comment by fmichaelobrien** [comment]
mike link to issues above for deletion

**Comment by fmichaelobrien** [comment]
#211 
#212

**Comment by obriensystems** [comment]
delete liens first

```
```
michael@cloudshell:~/wse_github/GoogleCloudPlatform (landing-zone-controller-e4g7d)$ gcloud config set project audit-prj-id-gz1
Updated property [core/project].
michael@cloudshell:~/wse_github/GoogleCloudPlatform (audit-prj-id-gz1)$ export AUDIT_LIEN=$(gcloud alpha resource-manager liens list --format json | jq .[0].name | sed 's/"//g' | sed 's/liens\///g')
michael@cloudshell:~/wse_github/GoogleCloudPlatform (audit-prj-id-gz1)$ echo $AUDIT_LIEN
p402270513653-lec39ae31-1d37-48f0-bbe7-59b0d41541a1
michael@cloudshell:~/wse_github/GoogleCloudPlatform (audit-prj-id-gz1)$ gcloud alpha resource-manager liens delete $AUDIT_LIEN 
Deleted [liens/p402270513653-lec39ae31-1d37-48f0-bbe7-59b0d41541a1].
michael@cloudshell:~/wse_github/GoogleCloudPlatform (audit-prj-id-gz1)$ gcloud config set project net-host-prj-prod-gz1
Updated property [core/project].
michael@cloudshell:~/wse_github/GoogleCloudPlatform (net-host-prj-prod-gz1)$ export AUDIT_LIEN=$(gcloud alpha resource-manager liens list --format json | jq .[0].name | sed 's/"//g' | sed 's/liens\///g')
michael@cloudshell:~/wse_github/GoogleCloudPlatform (net-host-prj-prod-gz1)$ echo $AUDIT_LIEN
p698859936700-l9cf73dda-4423-454a-b137-c838b6357574
michael@cloudshell:~/wse_github/GoogleCloudPlatform (net-host-prj-prod-gz1)$ gcloud alpha resource-manager liens delete $AUDIT_LIEN
Deleted [liens/p698859936700-l9cf73dda-4423-454a-b137-c838b6357574].

```
```
